probe/endpoint/connection_tracker.go

```diff
@@ -176,24 +173,152 @@ func (t *connectionTracker) getInitialState() {
-		var toNodeInfo, fromNodeInfo map[string]string
-		if e.pid > 0 {
-			fromNodeInfo = map[string]string{
-				process.PID:       strconv.Itoa(e.pid),
-				report.HostNodeID: hostNodeID,
+	/* Collect the connections by from/to address pairs (scoped by namespace) plus destination port
+	   There are three main cases:
+		    * connections from address+port off-box to a local process
+		      - in this case we know the pid of the local process
+		    * connections from local processes to an off-box address+port
+		      - we will know the pids of the local processes but not the remote
+		    * connections from local processes to a local process
+	          - these connections will each be reported twice by ebpf, as incoming and as outgoing.
+	*/
+	type triple struct {
+		fromAddr, toAddr [net.IPv4len]byte
+		networkNamespace uint32
+		toPort           uint16
+	}
+	connectionsByTriple := make(map[triple]mapPortToPids, 1000)
+	t.ebpfTracker.walkConnections(func(key ebpfKey, e ebpfDetail) {
+		var t triple
+		var fromPort uint16
+		if e.incoming {
+			t = triple{
+				fromAddr:         key.toAddr,
+				toAddr:           key.fromAddr,
+				toPort:           key.fromPort,
+				networkNamespace: key.networkNamespace,
+			}
+			fromPort = key.toPort
+		} else {
+			t = triple{
+				fromAddr:         key.fromAddr,
+				toAddr:           key.toAddr,
+				toPort:           key.toPort,
+				networkNamespace: key.networkNamespace,
 			}
+			fromPort = key.fromPort
+		}
+		portToPids := connectionsByTriple[t]
+		if portToPids == nil {
+			portToPids = make(mapPortToPids)
 		}
-		t.addConnection(rpt, e.incoming, e.tuple, e.networkNamespace, fromNodeInfo, toNodeInfo)
+		pids := portToPids[fromPort]
+		if e.incoming {
+			pids.toPid = e.pid
+		} else {
+			pids.fromPid = e.pid
+		}
+		portToPids[fromPort] = pids
+		connectionsByTriple[t] = portToPids
 	})
+
+	for triple, portToPids := range connectionsByTriple {
+		modulus, count := findModulus(portToPids)
+		i, sent, skipped := 0, 0, 0
+		// Now do the actual sends
+		for fromPort, pids := range portToPids {
+			if (fromPort % modulus) == 0 {
+				tuple := fourTuple{
+					fromAddr: triple.fromAddr,
+					fromPort: fromPort,
+					toAddr:   triple.toAddr,
+					toPort:   triple.toPort,
+				}
+				sent++
+				if sent == count {
+					skipped += (len(portToPids) - i - 1)
+				}
+				t.addConnection(rpt, hostNodeID, tuple, uint(pids.fromPid), uint(pids.toPid), triple.networkNamespace, skipped+1)
+				skipped = 0
+			} else {
+				skipped++
+			}
+			i++
+		}
+	}
+
 	return nil
 }
 
-func (t *connectionTracker) addConnection(rpt *report.Report, incoming bool, ft fourTuple, namespaceID uint32, extraFromNode, extraToNode map[string]string) {
-	if incoming {
-		ft = reverse(ft)
-		extraFromNode, extraToNode = extraToNode, extraFromNode
+// Pick a subset of the connections to send, such that if two probes
+// on different machines go through the same process there is a good
+// chance of overlap.
```

Comment:
If a process is continually opening and closing connections, then the originating port number will change over time, e.g. it could be 30002, 30004, 30006, ...  The connections will be open for a short time, then go into the "recently closed" set, then disappear.

When there are a big number, if we pick the first N then there's some chance someone else has observed 30062, 30064, 30066, ... and so the first N don't match.  If we pick modulus 3 or 9 or 27..., then a subset of ports from all across the numeric range will be sent, and there is more chance of a match.